src/main/java/com/enderio/base/common/item/tool/SoulVialItem.java

```diff
@@ -74,121 +91,119 @@ public void addDetailedTooltips(ItemStack itemStack, @Nullable Player player, Li
                     pPlayer.setItemInHand(pUsedHand, filledVial);
                 } else {
                     if (!pPlayer.addItem(filledVial)) {
                         pPlayer.drop(filledVial, false);
                     }
                 }
-
-                // Remove the captured mob.
-                pInteractionTarget.discard();
-            }
-
-            return InteractionResult.SUCCESS;
-        }).orElse(InteractionResult.SUCCESS);
+            },
+            component -> pPlayer.displayClientMessage(component, true)
+        );
     }
 
     // Release logic
-    @Nonnull
     @Override
     public InteractionResult useOn(UseOnContext pContext) {
         if (pContext.getLevel().isClientSide) {
             return InteractionResult.FAIL;
         }
 
-        ItemStack itemStack = pContext.getItemInHand();
         Player player = pContext.getPlayer();
 
         // Only players may use the soul vial
         if (player == null) {
             return InteractionResult.FAIL;
         }
+        return releaseEntity(pContext.getLevel(), pContext.getItemInHand(), pContext.getClickedFace(), pContext.getClickedPos(),
+            emptyVial -> player.setItemInHand(pContext.getHand(), emptyVial));
+    }
 
-        itemStack.getCapability(EIOCapabilities.ENTITY_STORAGE).ifPresent(entityStorage -> {
+
+    private static InteractionResult catchEntity(ItemStack soulVial, LivingEntity entity, Consumer<ItemStack> filledVialInsertion, Consumer<Component> displayCallback) {
+        return soulVial.getCapability(EIOCapabilities.ENTITY_STORAGE).map(entityStorage -> {
+            if (!entityStorage.hasStoredEntity()) {
+                // Don't allow bottled player.
+                if (entity instanceof Player) {
+                    displayCallback.accept(EIOLang.SOUL_VIAL_ERROR_PLAYER);
+                    return InteractionResult.FAIL;
+                }
+                // Get the entity type and verify it isn't blacklisted
+                switch (EntityCaptureUtils.getCapturableStatus(entity)) {
+                    case BOSS -> {
+                        displayCallback.accept(EIOLang.SOUL_VIAL_ERROR_PLAYER);
+                        return InteractionResult.FAIL;
+                    }
+                    case BLACKLISTED -> {
+                        displayCallback.accept(EIOLang.SOUL_VIAL_ERROR_BLACKLISTED);
+                        return InteractionResult.FAIL;
+                    }
+                    case INCOMPATIBLE -> {
+                        displayCallback.accept(EIOLang.SOUL_VIAL_ERROR_FAILED);
+                        return InteractionResult.FAIL;
+                    }
+                }
+
+                // No dead mobs.
+                if (!entity.isAlive()) {
+                    displayCallback.accept(EIOLang.SOUL_VIAL_ERROR_DEAD);
+                    return InteractionResult.FAIL;
+                }
```

Comment:
Could you review the spacing above these comments please? I feel its tidier if you place spaces above comments